---
title: "Project 03"
author: "Your Name"
date: "`r format(Sys.time(), '%Y%m%d')`"
output:
  github_document:
    html_preview: false
    toc: true
    toc_depth: 2
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
library('reticulate')
use_python('/usr/bin/python3')
```

---

# Gibbs sampling

---
## Imports

For this project, it tasks you with creating a sequence logo to plot the motif sequences identified through your algorithm.
This is a very minor portion of the project and is best used for checking your work. That means that if you cannot get the library
that you are about to install to work, don't worry. At a bare minimum, you only need `numpy` to complete this project.

```{r, install_seqlogo}
reticulate::conda_install(
  envname = "r-reticulate", 
  packages=c("seqlogo", "pdf2svg", "setuptools"),
  channel=c("bioconda", "conda-forge"), 
  conda="auto", 
  additionall_install_args="--solver=libmamba"
)
```


```{python}
import numpy as np
import random
#import seqlogo

#import function for building sequence motif & idenfitying seqs matching to motif
from data_readers import *
from seq_ops import get_seq
from motif_ops import *
```

---
## Implement Gibbs Sampler


Gibbs sampling is a MCMC approach to identify enrichments. Here we will implement a method to identify motifs from a set of regions. 

Important considerations:
- We will need to score each sequence with a PWM using the `score_kmer()` or `score_sequence()` functions
    - You will need to investigate into the help documenation and libraries to identify how best to use these functions. 
- These sites are often not strand-specific and so both scores on the negative as well as positive strand should be considered
- To select a random sequence, use `random.randint()` or `numpy.random.randint()`
- To select a new position $m$ (as defined below) use `random.choices()` or `numpy.random.choice()`  

Assumptions: 
- We know $k$ as the length of expected motif
- Each sequence contains the motif


```
GibbsMotifFinder(DNA, k-length)
    random pick of k-length sequences from each line of DNA as Motifs
    for j ← 1 to 10000 or Motifs stops changing
        i ← Random(N) where N is number of DNA entries
        PWM ← PWM constructed from all Motifs except for Motifi
        Motifi ← select position m from PWM-scored k-mers in DNAi in probabilistic fashion from score distribution
    return PFM
```

Probability of chosing position $m = \frac{A_{m}}{\sum_{l}A_{l}}$ for positions $l$ in DNAi


**Note:** I have also added a function to `motif_ops.py` that will calculate the information content of your motifs. This is useful to observe the progression of your Gibbs sampler as well as a measure of convergence. You can use this function as `IC = pfm_ic(pfm)`. You should expect a slow increase of IC until it plateaus such as in the plot below from your lecture slides:

![]('figures/Gibbs_Sampling.png')


```{python}
def GibbsMotifFinder(seqs, k, seed=None):
    '''
    Function to find a pfm from a list of strings using a Gibbs sampler
    
    Args: 
        seqs (str list): a list of sequences, not necessarily in same lengths
        k (int): the length of motif to find
        seed (int, default=42): seed for np.random

    Returns:
        pfm (numpy array): dimensions are 4xlength
        
    '''
    # Initialize RNG
    random.seed(seed)
    
    # For each sequence, initialize the best motif
    #   as a random substring of length k
    # Save the motifs and their indices to separate lists
    motifs = []
    motifs_i = []
    previous_IC = 0.0
    
    for seq in seqs:
        i = random.randint(0,len(seq)-k)
        motifs_i.append(i)
        motifs.append(seq[i:i+k])
    
    
    #print(motifs)
    #print(motifs_i)
    
    # Main loop
    for iterations in range(0,1000):
        # Randomly select a sequence
        seq_i = random.randint(0,len(seqs)-1)
        seq = seqs[seq_i]
        # Recalculate PWM using the motifs of all other sequences
        other_motifs = motifs[:seq_i] + motifs[seq_i+1:]
        freq_matrix = build_pfm(other_motifs, k)
        weight_matrix = build_pwm(freq_matrix)
        # Score each k-length sub-string of our chosen sequence with our PWM
        scores = score_substrings(seq, k, weight_matrix)
        # Update the chosen sequence's motif index probabilistically
        new_motif_i = random.choices(range(0,len(seq)-k+1), scores, k=1)
        print(new_motif_i)

        info_content = pfm_ic(freq_matrix)
        print(info_content)
        print(iterations)
        if abs(info_content - previous_IC) <= 0.000001:
            break
        
        previous_IC = info_content

    return freq_matrix


def score_substrings(seq, k, pwm):
    scores = []
    relative_scores = []
    
    for i in range(0,len(seq)-k+1):
        scores.append(score_kmer(seq[i:i+k],pwm))
        
    score_max = max(scores)
    score_min = min(scores)

    for score in scores:
        relative_scores.append((score - score_min)/(score_max - score_min))
        
    return relative_scores

```


```{python}
# Here we test your Gibbs sampler.
# You do not need to edit this or the section below. This is the Driver program

#read promoters, store in a list of strings
seq_file="data/GCF_000009045.1_ASM904v1_genomic.fna"
gff_file="data/GCF_000009045.1_ASM904v1_genomic.gff"

seqs = []

for name, seq in get_fasta(seq_file): # For each entry in our FASTA file
    for gff_entry in get_gff(gff_file): # For each entry in our GFF file
        if gff_entry.type == 'CDS': # If this is a coding sequence
            promoter_seq = get_seq(seq, gff_entry.start, gff_entry.end, gff_entry.strand, 50) # Extract 50 bp as a promoter
            if "AGGAGG" in promoter_seq:
                seqs.append(promoter_seq)
```


```{python}
# Run the gibbs sampler:
promoter_pfm = GibbsMotifFinder(seqs,10)
print(promoter_pfm)

# Plot the final pfm that is generated: 
#seqlogo.seqlogo(seqlogo.CompletePm(pfm = promoter_pfm.T))
```
